Vulkan: Graphics support does not work for multiple color attachments

Graphics support does not work for multiple color attachments. For example this amber script does not pass.

```groovy
SHADER vertex vert_shader PASSTHROUGH
SHADER fragment frag_shader GLSL
#version 430
layout(location = 0) out vec4 color_out;
layout(location = 1) out vec4 color_out1;
void main() {
  color_out = vec4(0.0, 0.5, 0.0, 0.8);
  color_out1 = vec4(1.0, 1.0, 1.0, 1.0);
}
END

BUFFER framebuffer FORMAT B8G8R8A8_UNORM
BUFFER framebuffer1 FORMAT B8G8R8A8_UNORM

PIPELINE graphics my_pipeline
  ATTACH vert_shader
  ATTACH frag_shader
  FRAMEBUFFER_SIZE 800 600
  BIND BUFFER framebuffer AS color LOCATION 0
  BIND BUFFER framebuffer1 AS color LOCATION 1
END

RUN my_pipeline DRAW_RECT POS 0 0 SIZE 250 250
EXPECT framebuffer IDX 0 0 SIZE 250 250 EQ_RGBA 0 127 0 204
EXPECT framebuffer1 IDX 0 0 SIZE 250 250 EQ_RGBA 255 255 255 255
```

```shell
[ERROR] validation layer (Validation):
 [ VUID-VkGraphicsPipelineCreateInfo-attachmentCount-00746 ] Object: 0x565254967b50 (Type = 3) | vkCreateGraphicsPipelines(): Render pass (0x6) subpass 0 has colorAttachmentCount of 2 which doesn't match the pColorBlendState->attachmentCount of 1. The Vulkan spec states: If rasterization is not disabled and the subpass uses color attachments, the attachmentCount member of pColorBlendState must be equal to the colorAttachmentCount used to create subpass (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-VkGraphicsPipelineCreateInfo-attachmentCount-00746)
../amber/tests/cases/draw_rect_mutliple_color_attachment.amber: Line 40: Probe failed at: 0, 0
  Expected RGBA: 255.000000, 255.000000, 255.000000, 255.000000
  Actual RGBA: 0.000000, 0.000000, 0.000000, 0.000000
Probe failed in 62500 pixels

Summary of Failures:
  ../amber/tests/cases/draw_rect_mutliple_color_attachment.amber

Summary: 0 pass, 1 fail
```

--- a/src/vulkan/graphics_pipeline.cc
+++ b/src/vulkan/graphics_pipeline.cc
@@ -422,11 +422,10 @@ Result GraphicsPipeline::CreateRenderPass() {
     ref.attachment = static_cast<uint32_t>(attachment_desc.size() - 1);
     ref.layout = VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL;
     color_refer.push_back(ref);
-
-    subpass_desc.colorAttachmentCount =
-        static_cast<uint32_t>(color_refer.size());
-    subpass_desc.pColorAttachments = color_refer.data();
   }
+  subpass_desc.colorAttachmentCount =
+        static_cast<uint32_t>(color_refer.size());
+  subpass_desc.pColorAttachments = color_refer.data();
 
   if (depth_stencil_format_.IsFormatKnown()) {
     attachment_desc.push_back(kDefaultAttachmentDesc);
@@ -503,26 +502,31 @@ GraphicsPipeline::GetVkPipelineDepthStencilInfo(
   return depthstencil_info;
 }
 
-VkPipelineColorBlendAttachmentState
+std::vector<VkPipelineColorBlendAttachmentState>
 GraphicsPipeline::GetVkPipelineColorBlendAttachmentState(
     const PipelineData* pipeline_data) {
-  VkPipelineColorBlendAttachmentState colorblend_attachment =
-      VkPipelineColorBlendAttachmentState();
-  colorblend_attachment.blendEnable = pipeline_data->GetEnableBlend();
-  colorblend_attachment.srcColorBlendFactor =
-      ToVkBlendFactor(pipeline_data->GetSrcColorBlendFactor());
-  colorblend_attachment.dstColorBlendFactor =
-      ToVkBlendFactor(pipeline_data->GetDstColorBlendFactor());
-  colorblend_attachment.colorBlendOp =
-      ToVkBlendOp(pipeline_data->GetColorBlendOp());
-  colorblend_attachment.srcAlphaBlendFactor =
-      ToVkBlendFactor(pipeline_data->GetSrcAlphaBlendFactor());
-  colorblend_attachment.dstAlphaBlendFactor =
-      ToVkBlendFactor(pipeline_data->GetDstAlphaBlendFactor());
-  colorblend_attachment.alphaBlendOp =
-      ToVkBlendOp(pipeline_data->GetAlphaBlendOp());
-  colorblend_attachment.colorWriteMask = pipeline_data->GetColorWriteMask();
-  return colorblend_attachment;
+  std::vector<VkPipelineColorBlendAttachmentState> states;
+
+  for (size_t i = 0; i < color_buffers_.size(); ++i) {
+    VkPipelineColorBlendAttachmentState colorblend_attachment =
+        VkPipelineColorBlendAttachmentState();
+    colorblend_attachment.blendEnable = pipeline_data->GetEnableBlend();
+    colorblend_attachment.srcColorBlendFactor =
+        ToVkBlendFactor(pipeline_data->GetSrcColorBlendFactor());
+    colorblend_attachment.dstColorBlendFactor =
+        ToVkBlendFactor(pipeline_data->GetDstColorBlendFactor());
+    colorblend_attachment.colorBlendOp =
+        ToVkBlendOp(pipeline_data->GetColorBlendOp());
+    colorblend_attachment.srcAlphaBlendFactor =
+        ToVkBlendFactor(pipeline_data->GetSrcAlphaBlendFactor());
+    colorblend_attachment.dstAlphaBlendFactor =
+        ToVkBlendFactor(pipeline_data->GetDstAlphaBlendFactor());
+    colorblend_attachment.alphaBlendOp =
+        ToVkBlendOp(pipeline_data->GetAlphaBlendOp());
+    colorblend_attachment.colorWriteMask = pipeline_data->GetColorWriteMask();
+    states.push_back(colorblend_attachment);
+  }
+  return states;
 }
 
 Result GraphicsPipeline::CreateVkGraphicsPipeline(
@@ -653,16 +657,16 @@ Result GraphicsPipeline::CreateVkGraphicsPipeline(
 
   VkPipelineColorBlendStateCreateInfo colorblend_info =
       VkPipelineColorBlendStateCreateInfo();
-  VkPipelineColorBlendAttachmentState colorblend_attachment;
 
-  colorblend_attachment = GetVkPipelineColorBlendAttachmentState(pipeline_data);
+  auto colorblend_attachment = GetVkPipelineColorBlendAttachmentState(pipeline_data);
 
   colorblend_info.sType =
       VK_STRUCTURE_TYPE_PIPELINE_COLOR_BLEND_STATE_CREATE_INFO;
   colorblend_info.logicOpEnable = pipeline_data->GetEnableLogicOp();
   colorblend_info.logicOp = ToVkLogicOp(pipeline_data->GetLogicOp());
-  colorblend_info.attachmentCount = 1;
-  colorblend_info.pAttachments = &colorblend_attachment;
+  colorblend_info.attachmentCount =
+      static_cast<uint32_t>(colorblend_attachment.size());
+  colorblend_info.pAttachments = colorblend_attachment.data();
   pipeline_info.pColorBlendState = &colorblend_info;
 
   pipeline_info.layout = pipeline_layout;

--- a/src/vulkan/graphics_pipeline.cc
+++ b/src/vulkan/graphics_pipeline.cc
@@ -423,8 +423,7 @@ Result GraphicsPipeline::CreateRenderPass() {
     ref.layout = VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL;
     color_refer.push_back(ref);
   }
-  subpass_desc.colorAttachmentCount =
-        static_cast<uint32_t>(color_refer.size());
+  subpass_desc.colorAttachmentCount = static_cast<uint32_t>(color_refer.size());
   subpass_desc.pColorAttachments = color_refer.data();
 
   if (depth_stencil_format_.IsFormatKnown()) {

--- a/src/vulkan/graphics_pipeline.cc
+++ b/src/vulkan/graphics_pipeline.cc
@@ -658,7 +658,8 @@ Result GraphicsPipeline::CreateVkGraphicsPipeline(
   VkPipelineColorBlendStateCreateInfo colorblend_info =
       VkPipelineColorBlendStateCreateInfo();
 
-  auto colorblend_attachment = GetVkPipelineColorBlendAttachmentState(pipeline_data);
+  auto colorblend_attachment =
+      GetVkPipelineColorBlendAttachmentState(pipeline_data);
 
   colorblend_info.sType =
       VK_STRUCTURE_TYPE_PIPELINE_COLOR_BLEND_STATE_CREATE_INFO;

--- a/src/vulkan/graphics_pipeline.h
+++ b/src/vulkan/graphics_pipeline.h
@@ -83,7 +83,7 @@ class GraphicsPipeline : public Pipeline {
   VkPipelineDepthStencilStateCreateInfo GetVkPipelineDepthStencilInfo(
       const PipelineData* pipeline_data);
   std::vector<VkPipelineColorBlendAttachmentState>
-      GetVkPipelineColorBlendAttachmentState(const PipelineData* pipeline_data);
+  GetVkPipelineColorBlendAttachmentState(const PipelineData* pipeline_data);
 
   VkRenderPass render_pass_ = VK_NULL_HANDLE;
   std::unique_ptr<FrameBuffer> frame_;

--- a/src/vulkan/graphics_pipeline.h
+++ b/src/vulkan/graphics_pipeline.h
@@ -82,8 +82,8 @@ class GraphicsPipeline : public Pipeline {
 
   VkPipelineDepthStencilStateCreateInfo GetVkPipelineDepthStencilInfo(
       const PipelineData* pipeline_data);
-  VkPipelineColorBlendAttachmentState GetVkPipelineColorBlendAttachmentState(
-      const PipelineData* pipeline_data);
+  std::vector<VkPipelineColorBlendAttachmentState>
+      GetVkPipelineColorBlendAttachmentState(const PipelineData* pipeline_data);
 
   VkRenderPass render_pass_ = VK_NULL_HANDLE;
   std::unique_ptr<FrameBuffer> frame_;

